hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/updatetokencustomfees/UpdateTokenCustomFeesDecoder.java

```diff
@@ -0,0 +1,201 @@
+
+    // Fractional Fee tuple indexes
+    private static final int FRACTIONAL_FEE_NUMERATOR = 0;
+    private static final int FRACTIONAL_FEE_DENOMINATOR = 1;
+    private static final int FRACTIONAL_FEE_MIN_AMOUNT = 2;
+    private static final int FRACTIONAL_FEE_MAX_AMOUNT = 3;
+    private static final int FRACTIONAL_FEE_NET_OF_TRANSFERS = 4;
+    private static final int FRACTIONAL_FEE_FEE_COLLECTOR = 5;
+
+    // Royalty Fee tuple indexes
+    private static final int ROYALTY_FEE_NUMERATOR = 0;
+    private static final int ROYALTY_FEE_DENOMINATOR = 1;
+    private static final int ROYALTY_FEE_AMOUNT = 2;
+    private static final int ROYALTY_FEE_TOKEN_ID = 3;
+    private static final int ROYALTY_FEE_USE_HBARS_FOR_PAYMENT = 4;
+    private static final int ROYALTY_FEE_FEE_COLLECTOR = 5;
+
+    @Inject
+    public UpdateTokenCustomFeesDecoder() {}
+
+    public @Nullable TransactionBody decodeUpdateFungibleTokenCustomFees(@NonNull final HtsCallAttempt attempt) {
+        final var call = UpdateTokenCustomFeesTranslator.UPDATE_FUNGIBLE_TOKEN_CUSTOM_FEES_FUNCTION.decodeCall(
+                attempt.inputBytes());
+        return TransactionBody.newBuilder()
+                .tokenFeeScheduleUpdate(
+                        updateTokenCustomFees(call, attempt.addressIdConverter(), this::customFungibleFees))
+                .build();
+    }
+
+    public TransactionBody decodeUpdateNonFungibleTokenCustomFees(@NonNull final HtsCallAttempt attempt) {
+        final var call = UpdateTokenCustomFeesTranslator.UPDATE_NON_FUNGIBLE_TOKEN_CUSTOM_FEES_FUNCTION.decodeCall(
+                attempt.inputBytes());
+        return TransactionBody.newBuilder()
+                .tokenFeeScheduleUpdate(
+                        updateTokenCustomFees(call, attempt.addressIdConverter(), this::customNonFungibleFees))
+                .build();
+    }
+
+    private TokenFeeScheduleUpdateTransactionBody updateTokenCustomFees(
+            @NonNull final Tuple call,
+            @NonNull final AddressIdConverter addressIdConverter,
+            @NonNull final TriFunction<Tuple, AddressIdConverter, TokenID, List<CustomFee>> customFees) {
+        final Address tokenAddress = call.get(TOKEN_ADDRESS);
+        final var tokenId = ConversionUtils.asTokenId(tokenAddress);
+        return TokenFeeScheduleUpdateTransactionBody.newBuilder()
+                .tokenId(tokenId)
+                .customFees(customFees.apply(call, addressIdConverter, tokenId))
+                .build();
+    }
+
+    private List<CustomFee> customFungibleFees(
+            @NonNull Tuple call, @NonNull final AddressIdConverter addressIdConverter, @NonNull final TokenID tokenId) {
+        return Stream.concat(
+                        decodeFixedFees(call.get(FIXED_FEE), addressIdConverter, tokenId),
+                        decodeFractionalFees(call.get(FRACTIONAL_FEE), addressIdConverter))
+                .toList();
+    }
+
+    private List<CustomFee> customNonFungibleFees(
```

Comment:
YMMV but I personally find these methods that `transform` a collection of things into a collection of different things easy to read.  Prefer them to for loops.  Combining streams + builders is pretty effective.  

(There's a little bit of syntax noise but it isn't bad.)

Reply:
I totally agree